Config files are now loaded via enviroment variables.

nickrobison-usds opened the pull request:
First off, thanks for your work on a great library! This is a proposed set of changes that modifies the way config files and loaded and merged at runtime.

Starting the server with no config arguments causes the application to look for an `application.local.conf` file in the current working directory, followed any application files in the resources folder. In addition, it adds any command line and environment variables to the config, with a higher priority than the config files.

Existing behavior is still preserved when manually specifying a config file, no auto-loading or config merging is performed.

This only a partial PR, tests still need to be written, but the hope is to gauge your interest in accepting these changes. I've also modified the POM to fix some compilation issues under Java 11. Thanks!

nickrobison-usds committed:
Config files are now loaded via enviroment variables.

Starting the server with no config arguments causes the application to look for an `application.local.conf` file in the current working directory, followed any application files in the resources folder. In addition, it adds any command line and environment variables to the config, with a higher priority than the config files.

The legacy behavior is still preserved when manually specifying a config file, no auto-loading or config merging is performed.

mestevens commented:
To be honest with you, it's been a minute since I've actively developed this repo, so I am open to PRs that come my way on it. That being said, maybe you can explain a bit about what the benefit of this PR is (I think I may just be missing the point)?

To my understanding what you're trying to do (when it's finished) is have the config check an environment variable to see what directory it should be loading the configuration from. Is this not something that this already supports by just passing the environment variable on startup?

For example `java -jar server-name.jar server $CONFIG_FILE` should read the config file wherever the environment variable points to.

Again, I haven't worked on this for a while, so I feel like I'm probably missing the point, so some clarification would be appreciated.

nickrobison-usds committed:
Fix null pointer when environment variable doesn’t exist

If the user did not set the ENV variable, a null pointer exception would be generated.

This fixes that error by using a safe way to check if the value is present, before trying to dereference it.

nickrobison-usds committed:
Improved overload resolution and added unit tests

The previous implementation wasn’t correctly updating the config object, which has been fixed.

Added some simple unit tests to verify that the overload and merging logic works as intended.

nickrobison-usds committed:
Updated environment key and override orderings

The environment override key is now uppercase, to match best practices.
It’s also now correctly overriden either by the environment variable, or the system property.

nickrobison-usds committed:
Invalidate Config caches in order to pickup latest property values

nickrobison-usds committed:
Moved property invalidation into the build method

When overriding during testing, the overrides are not applied when the constructor is called, so we have to move them down into the build method itself.
